### ISSUE TITLE
stream from any audio source and players

### ISSUE THREAD

**Issue by jbonlinea** [badaix/snapcast#219]
Hi there, 

I discovered `Snapcast` recently and it seems awesome ! congrats ! 

On you home page you explain that the aim is to set up a chain like this 

> audio player software -> snapfifo -> snapserver -> network -> snapclient -> alsa

Ok, I get the idea and I've seen quite many tutorial on how to set up a _multiroom sonos like system_, when the _audio player software_ is an alternative to `Music Player Deamon` (MDP) like `Mopidy`, `Volumio` or `Rune audio`.

Great, but that's already one step ahead too far. 
Indeed, all these audio player software are "big things", and **I simply would like to stream (all) the sound of my Raspberry pi**, I mean exactly what come out through HDMI, wether it comes from a website I'm visiting, a video or audio file I'm playing in VLC... 
I'm kind of a noobs but it seems to me as the starting point.

Then, I may add that : 
- the _audio playing software_ might be a _media center_ like `Kodi` (installed on top of raspbian, not a dedicated distribution). To my understanding if the above is acheived, then this would work naturally.
- the **audio output** might be a _DAC hat_ (like the Hifiberry Dac+) or a _USB soundcard_.
- the **audio source is distant,** for instance played on a android/ios terminal. I've seen one example that use `Shairport-sync` but the explaination are a bit rudimentary, [see](https://nik.re/2016/09/04/mopidy-shairport-sync-and-snapcast-is-a-gread-combo). What about bluetooth, turning your Rpi as a` bluetooth A2CP receiver` [like here](https://github.com/BaReinhard/Super-Simple-Raspberry-Pi-Audio-Receiver-Install)  which would stream the audio.

Before going multiroom, I would like to have one device (an Rpi) which is `Snapcast server` and `Snapcast client` at the same time (I've seen this acheived), and whitch accept all the above mentioned audio sources, audio players, and audio outputs.

Only then I may :
- add Snapcast client for multiroom
- go for zoning whith various client streaming different audio source

I've read quite a bit already, but this never seems addressed ! 

So basically the question are how to set up Snapcast in order that 
- it may streams every audio source (local app, bluetooth, shaireport, etc.) ?
- you may direct various _audio source_, or varuois _audio player software_ toward the clients you wish .

Is there somlething wrong in my thinking or may you give me some orientations.

Regards

**Comment by palitu** [comment]
I think that this may work for you:

https://github.com/badaix/snapcast/issues/208

I have a setup which routes all my audio playing on a machine to snapserver.  It initialises pulse audio to default to the snapserver fifo file.  If you want to have the audio played on that machine, you need to start the snapclient on it as well.

so the flow is something like this:
audio source (Spotify, Browser, Kodi, etc) > Pulse Audio > snapfifo > snapserver > snapclients (localhost, client 1 .. client n).


**NOTE**
The one issue i cannot see being resolved is audio from videos (like from kodi).  there is a small delay (0.5s or so) playing to snapcast.  this WILL NOT work for playing video audio, until someone creates a player which will sync the video time and the audio time.

**Comment by palitu** [comment]
Also i am trying to avoid MPD, mopidy, rune etc to avoid the added complexity that it results in.

**Comment by badaix** [comment]
the delay depends on your configuration. The default buffer is actually set to 1s. But you can lower it with as server parameter to 500ms, which should still work fine.

**Comment by jbonlinea** [comment]
Hi there :-)

> Also i am trying to avoid MPD, mopidy, rune etc to avoid the added complexity that it results in.

Ok so It seem my approach make sense !

> I think that this may work for you: #208

Excelent !


> I have a setup which routes all my audio playing on a machine to snapserver. It initialises pulse audio to default to the snapserver fifo file.

Spot on !
first things first ;-)

>  If you want to have the audio played on that machine, you need to start the snapclient on it as well.

yes yes ! an it actually seem flawless ! 

> so the flow is something like this:
audio source (Spotify, Browser, Kodi, etc) > Pulse Audio > snapfifo > snapserver > snapclients (localhost, client 1 .. client n).

Exactly, what I was thinkig of !
Especially since I belive `Pulse audio` can also be used to _tunel_ `Bluetooth A2DP` and _Airplay_ based on `Shairport` or `Shairport sync`

> NOTE
The one issue i cannot see being resolved is audio from videos (like from kodi). there is a small delay (0.5s or so) playing to snapcast. this WILL NOT work for playing video audio, until someone creates a player which will sync the video time and the audio time.

Well hey, Kodi has a GUI, on the fly _function to delay or advance audio_ ! So if Snapcast delay audio by a 1s buffer to secure continuous streaming (thank's @badaix), the we may advance audio by the same amount in Kodi ! Problem solved.
And with a bit of luck it might even be set up as default ! 

And you know what, VLC also have such function, see [here](http://lifehacker.com/367558/fix-desynchronized-video-and-audio-with-vlc) and [here](http://lifehacker.com/5910943/fix-out-of-sync-audio-in-vlc-with-a-keyboard-shortcut).

So Between VLC, Kodi and its bunch of plugins, we've already covered a nice share of usage, if not most !
Ok I do belive that being able to sync audio and video from a web browser would be a nice improvement, but as we say in french " the best is the ennemy of the good"

And this is already pretty darn good !

Excelent work guys ! 

@palitu : if you could provide a detailled tuto of your set up,  i'll try to integrate it with [this](https://github.com/BaReinhard/Super-Simple-Raspberry-Pi-Audio-Receiver-Install) in order to set up the most featured and flexible audio receiver and multiroom system;

Thank's

**Comment by palitu** [comment]
@jbonlinea, i'm not sure how much extra i can give you in terms of a tutorial.  The steps are pretty much the same as the linked system above, and i have only got decent steps for setting up the server.  I am still waiting for my rPi Zero-W to arrive, with a little amp hat, so i cannot give details on the client side.

**Server**:
1. Install your media player (Spotify, Kodi)
2. give your current user access to modify Pulse Audio settings
3. Modify pulse audio settings file to create the snapcast audio sink and make it the default
4. Modify snapcast settings to ensure that it 44.1k and doesnt try to create the stream it self.
5. Restart server so that the audio sets itself up correctly

**Client** (and this is best guess for a rPi):
1. install your operating system (can be headless)
2. connect to the internet
3. download and install the appropriate ARM release of snapclient
4. Set the name of your client (i know you can, just dont know how)
5. restart!

then you should have the server and first node set up.

**Comment by badaix** [comment]
Sometimes I also have to restart snapclient after installation. 
Doing a `sudo dpkg --purge snapclient` may help to clean up everything before the installation of the new version. Especially from version 0.10 to 0.11 some paths have changed.

**Comment by palitu** [comment]
Updated the client instructions for a rPi-W, with the X400 DAC/Amp.

love it!
